x-pack/apm-server/aggregation/servicemetrics/aggregator.go

```diff
@@ -0,0 +1,342 @@
+		sum:         duration,
+		min:         duration,
+		max:         duration,
+		sumFailures: 0,
+		agentName:   event.Agent.Name,
+	}
+	if a.active.storeOrUpdate(key, metrics, event, a.config.Logger) {
+		return model.APMEvent{}
+	}
+	return makeMetricset(key, metrics)
+}
+
+type metricsBuffer struct {
+	maxSize int
+
+	mu sync.RWMutex
+	m  map[aggregationKey]serviceMetrics
+}
+
+func newMetricsBuffer(maxSize int) *metricsBuffer {
+	return &metricsBuffer{
+		maxSize: maxSize,
+		m:       make(map[aggregationKey]serviceMetrics),
+	}
+}
+
+func (mb *metricsBuffer) storeOrUpdate(key aggregationKey, value serviceMetrics, event *model.APMEvent, logger *logp.Logger) bool {
+	mb.mu.Lock()
+	defer mb.mu.Unlock()
+	old, ok := mb.m[key]
+	if !ok {
+		n := len(mb.m)
+		half := mb.maxSize / 2
+		if !ok {
+			switch n {
+			case mb.maxSize:
+				return false
+			case half - 1:
+				logger.Warn("service metrics groups reached 50% capacity")
+			case mb.maxSize - 1:
+				logger.Warn("service metrics groups reached 100% capacity")
+			}
+		}
+	}
+	var max = old.max
+	var min = old.min
+	if value.max > max {
+		max = value.max
+	}
+	if value.min < min {
+		min = value.min
+	}
+	var sumNewFailures = 0
+	if event.Event.Outcome == "failure" {
+		sumNewFailures = 1
+	}
+	var agentName = old.agentName
+	if old.agentName == "" {
+		agentName = event.Agent.Name
+	}
+	var metrics = serviceMetrics{
+		count:       value.count + old.count,
+		sum:         value.sum + old.sum,
+		max:         max,
+		min:         min,
+		sumFailures: old.sumFailures + sumNewFailures,
+		agentName:   agentName,
+	}
+	mb.m[key] = metrics
+	return true
+}
+
+type aggregationKey struct {
+	timestamp time.Time
+
+	// origin
+	serviceName        string
+	serviceEnvironment string
+	transactionType    string
+
+	outcome string
+}
+
+func makeAggregationKey(event *model.APMEvent, interval time.Duration) aggregationKey {
+	var transactionType = ""
+	if event.Transaction != nil {
+		transactionType = event.Transaction.Type
+	} else if event.Span != nil {
+		transactionType = event.Span.Type
+	}
+	return aggregationKey{
+		// Group metrics by time interval.
+		timestamp: event.Timestamp.Truncate(interval),
+
+		serviceName:        event.Service.Name,
+		serviceEnvironment: event.Service.Environment,
+		transactionType:    transactionType,
+	}
+}
+
+type serviceMetrics struct {
+	count int64
+	sum   int64
+
+	min int64
+	max int64
+
+	sumFailures int
+	agentName   string
+}
+
+func makeMetricset(key aggregationKey, metrics serviceMetrics) model.APMEvent {
+	return model.APMEvent{
+		Timestamp: key.timestamp,
+		Service: model.Service{
+			Name:        key.serviceName,
+			Environment: key.serviceEnvironment,
+		},
+		Event: model.Event{
+			Outcome: key.outcome,
+		},
+		Agent: model.Agent{
+			Name: metrics.agentName,
+		},
+		Processor: model.MetricsetProcessor,
+		Metricset: &model.Metricset{
+			Name: metricsetName,
+		},
+		Transaction: &model.Transaction{
+			FailureCount: &metrics.sumFailures,
```

Comment:
Missing transaction type in emitted doc (already part of aggregation key)

Reply:
Why do we need to aggregate on transaction type? If we're aggregating on transaction type, then they're starting to look a lot like transaction metrics, not service metrics.